lib/galaxy/managers/notification.py

```diff
@@ -110,31 +151,118 @@ def ensure_notifications_enabled(self):
         if not self.notifications_enabled:
             raise ConfigDoesNotAllowException("Notifications are disabled in this Galaxy.")
 
+    @property
+    def can_send_notifications_async(self):
+        return self.config.enable_celery_tasks
+
     def send_notification_to_recipients(self, request: NotificationCreateRequest) -> Tuple[Optional[Notification], int]:
         """
         Creates a new notification and associates it with all the recipient users.
+
+        It takes into account the user's notification preferences to decide if the notification should be sent to them.
+        No other notification channel is used here, only the internal database associations are created.
         """
         self.ensure_notifications_enabled()
         recipient_users = self.recipient_resolver.resolve(request.recipients)
-        notifications_sent = len(recipient_users)
-        notification = self._create_notification_model(request.notification)
+        notification = self._create_notification_model(request.notification, request.galaxy_url)
         self.sa_session.add(notification)
-        self._send_to_users(notification, recipient_users)
         with transaction(self.sa_session):
             self.sa_session.commit()
+
+        notifications_sent = self._create_associations(notification, recipient_users)
+        with transaction(self.sa_session):
+            self.sa_session.commit()
+
         return notification, notifications_sent
 
-    def _send_to_users(self, notification: Notification, users: List[User]):
-        # TODO: Move this potentially expensive operation to a task?
+    def _create_associations(self, notification: Notification, users: List[User]) -> int:
+        success_count = 0
+        for user in users:
+            try:
+                if self._is_user_subscribed_to_category(user, notification.category):  # type:ignore[arg-type]
+                    user_notification_association = UserNotificationAssociation(user, notification)
+                    self.sa_session.add(user_notification_association)
+                    success_count += 1
+            except Exception as e:
+                log.error(f"Error sending notification to user {user.id}. Reason: {unicodify(e)}")
+                continue
+        return success_count
+
+    def dispatch_pending_notifications_via_channels(self) -> int:
+        """
+        Dispatches all pending notifications to the users depending on the configured channels.
+
+        This is meant to be called periodically by a background task.
+        """
+        self.ensure_notifications_enabled()
+        pending_notifications = self.get_pending_notifications()
+
+        # Mark all pending notifications as dispatched
+        for notification in pending_notifications:
+            notification.dispatched = True
```

Comment:
but what if an error happens in the next loop and a notification is not dispatched via `_dispatch_notification_to_users`? Then its state in the database would be incorrect, right?

Reply:
I thought a bit about that, and I chose not to enforce a successful dispatch to all users (maybe using retries) because it feels "less annoying" from the user's perspective to receive no email than to receive it potentially multiple times...

Thinking about it again now (thanks to your comment :laughing:), maybe it would be better to move the dispatched column to the UserNotificationAssociation and have more fine control over which users the notification was "dispatched" successfully and for which it failed :thinking: Then we could think how to handle the second try for those that failed... What do you think?

Reply:
It makes sense, I think. The downside is we no longer have a quick way to determine if a notification has been dispatched (checking status of each notification based on all existing user associations is a no-go: performance + no easy way to handle new users who would qualify for a given notification, but where created after the notification was sent - it gets hairy..). We could have a status field on the Notification, set only when all users have been processed. However, if "has been processed" is defined as "task completed without error" then it's safe, I think, since it's dependent only on our code not raising an error. But if it's any more detailed (has been sent or has been received, etc.), we may end up with lots of "pending" notifications because of a problem dispatching to one user out of thousands.

I think the main question is how do we want to handle a failed notification dispatch. Sorry, maybe it's simpler than I think and all these are non-issues..

Reply:
Then, probably the current approach is safe enough. This is what happens now:
### Phase 1: Sending the notification
1. An admin sends a notification to X users, Y groups, and/or Z roles
2. The `Notification` is created with `dispatched=false`
3. A `UserNotificationAssociation` is created for each user after resolving X, Y, and Z into a flat set of users (only those who didn't opt out from that notification)
4. At this point, the users can see their notifications in Galaxy (unless the notification was scheduled)

### Phase 2: Dispatching notifications to other communication channels
5. A periodic task (every 10 minutes by default) tries to dispatch all "pending" notifications (dispatched=False)
6. The task grabs the currently pending notifications and sets them as dispatched
7. For each notification
8. For each associated user
9. Send a notification through every channel (currently only send an email)

The idea behind `step 6`, marking them as dispatched before processing, is to avoid a second task grabbing some of those notifications and duplicating emails (etc.) if, for some reason, sending the emails takes longer than the periodic check.

If sending an email fails, it will not affect other users, but there will be no retry and the email will be lost. I think that is preferable over getting multiple potential emails.

Reply:
Yes, I think we can leave as is.  I think, it makes sense especially given items 1.4 and 2.9: if a notificaton is displayed to the user (1.4) it feels sufficient to refer to it as dispatched; and, given that we may have multiple channels, tracking successful dispatches for each channel is a whole lot of additional logic I hope we'll never have to implement in the context of Galaxy.. 

Thank you for the detailed description!